common/client/multi_node.go

```diff
@@ -40,165 +37,160 @@ var (
-	classifySendTxError func(tx TX, err error) SendTxReturnCode,
-	sendTxSoftTimeout time.Duration,
-) MultiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT, BATCH_ELEM] {
-	nodeSelector := newNodeSelector(selectionMode, nodes)
+	selectionMode string, // type of the "best" RPC selector (e.g HighestHead, RoundRobin, etc.)
+	leaseDuration time.Duration, // defines interval on which new "best" RPC should be selected
+	primaryNodes []Node[CHAIN_ID, HEAD, RPC_CLIENT],
+	sendOnlyNodes []SendOnlyNode[CHAIN_ID, RPC_CLIENT],
+	chainID CHAIN_ID, // configured chain ID (used to verify that passed primaryNodes belong to the same chain)
+	chainFamily string, // name of the chain family - used in the metrics
+) MultiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT] {
+	// TODO: does node selector only need primary nodes, or all nodes?
+	nodeSelector := newNodeSelector(selectionMode, primaryNodes)
 	// Prometheus' default interval is 15s, set this to under 7.5s to avoid
 	// aliasing (see: https://en.wikipedia.org/wiki/Nyquist_frequency)
 	const reportInterval = 6500 * time.Millisecond
-	if sendTxSoftTimeout == 0 {
-		sendTxSoftTimeout = QueryTimeout / 2
-	}
-	c := &multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT, BATCH_ELEM]{
-		nodes:               nodes,
-		sendonlys:           sendonlys,
-		chainID:             chainID,
-		chainType:           chainType,
-		lggr:                logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
-		selectionMode:       selectionMode,
-		noNewHeadsThreshold: noNewHeadsThreshold,
-		nodeSelector:        nodeSelector,
-		chStop:              make(services.StopChan),
-		leaseDuration:       leaseDuration,
-		chainFamily:         chainFamily,
-		classifySendTxError: classifySendTxError,
-		reportInterval:      reportInterval,
-		sendTxSoftTimeout:   sendTxSoftTimeout,
+	c := &multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]{
+		primaryNodes:   primaryNodes,
+		sendOnlyNodes:  sendOnlyNodes,
+		chainID:        chainID,
+		lggr:           logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
+		selectionMode:  selectionMode,
+		nodeSelector:   nodeSelector,
+		chStop:         make(services.StopChan),
+		leaseDuration:  leaseDuration,
+		chainFamily:    chainFamily,
+		reportInterval: reportInterval,
 	}
 
 	c.lggr.Debugf("The MultiNode is configured to use NodeSelectionMode: %s", selectionMode)
 
 	return c
 }
 
+func (c *multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]) ChainType() config.ChainType {
+	return c.chainType
+}
+
+func (c *multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]) DoAll(ctx context.Context, do func(ctx context.Context, rpc RPC_CLIENT, isSendOnly bool) bool) error {
+	callsCompleted := 0
+	for _, n := range c.primaryNodes {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if n.State() != nodeStateAlive {
+			continue
+		}
+		if do(ctx, n.RPC(), false) {
```

Comment:
Why would we want to abort all calls in this case? This could also happen after a few calls already returned true, in which case why shouldn't we complete as many calls as possible?